### DISCUSSION TITLE
Optimizing motion implementation

### DISCUSSION THREAD

**Discussion by GrumpyMeow** [blakeblackshear/frigate#11456] [General]
I think i see an opportunity to improve the performance "improved_motion"-implementation by replacing "scipy.ndimage.gaussian_filter" with "cv2.GaussianBlur".  Is there maybe a specific reason why this optimization isn't implemented at this time?   

I think i see a 10x performance improvement by using the cv2-version. 

The original:
```
resized_frame = gaussian_filter(resized_frame, sigma=1, radius=self.blur_radius)
```
Using  cv2:
```
resized_frame = cv2.GaussianBlur(resized_frame, (self.blur_radius, self.blur_radius), 0)
```
I'll continue to see if the above indeed works as intended...

**Comment by NickM-27** [comment]
cv2 was used before during the motion detection testing and some users saw massive CPU usage increase due to some underlying python-cv2 issues. 

You can see

https://github.com/blakeblackshear/frigate/issues/6853

and

https://github.com/blakeblackshear/frigate/pull/6940

for context. Most likely in a future version of frigate motion detection will be rewritten in a different language and avoid using python altogether

**Reply by NickM-27** [reply]
There is a lot there, https://github.com/blakeblackshear/frigate/issues/6853#issuecomment-1605205491 specifically

> I spent some time trying to back out specific parts of the new motion detector to see what was driving the change in CPU usage.
> 
> I found that if I removed both the gaussian blur and contrast improvement calls, the CPU usage dropped back to very low levels.
> 
> What is interesting is that adding either of them increases CPU usage to the elevated levels. Adding a second one doesn't seem to increase levels further. It doesn't matter which one is enabled, just that at least one is.
> 
> I'm wondering if there is some underlying overhead associated with the opencv interface to python that initializes once for a given scope. I did try updating to the latest opencv and it reduced the CPU usage by half, but it's still elevated with either of those function calls enabled.
> 
> Just some weird python shit. One day I want to port this part of the code base to rust or something.
> 
> Another idea would be to look at doing the same operations in numpy and optimizing with numba, but there is a bunch of other opencv stuff in there.
